databuilder/__init__.py

```diff
@@ -1,3 +1,7 @@
 from pathlib import Path
 
 __version__ = Path(__file__).parent.joinpath("VERSION").read_text().strip()
+
+# This is populated in databuilder.__main__.main with a list of any command line
+# arguments that follow "--", and can be used to parameterise a dataset definition.
+user_args = None
```

Comment:
I think it would be useful to maintain the property that dataset definitions are executable without error directly in Python i.e. by running
```sh
python analyis/dataset_definition.py
```

As that doesn't go through `databuilder.__main__` it won't end up setting `user_args`.

I think it would be reasonable to populate `user_args` at import time, given that `sys.argv` is. But maybe it would be cleaner to make it a function like `get_user_args()` e.g.
```py
def split_cmd_and_user_args(args):
    if args is None:
        args = sys.argv[1:]
    if "--" in args:
        separator_index = args.index("--")
        cmd_args = args[:separator_index]
        user_args = args[separator_index + 1:]
    else:
        cmd_args = args
        user_args = []
    return cmd_args, user_args


def get_cmd_args(args=None):
    return split_cmd_and_user_args(args)[0]


def get_user_args(args=None):
    return split_cmd_and_user_args(args)[1]
```

Reply:
Thanks.  Would setting `user_args = []` in databuilder/__init__.py do the trick?

Reply:
It would mean you had no way of passing arguments to the dataset definition other than by invoking it via databuilder. I think it would be nice if you could do:
```
python analysis/dataset_definition.py -- foo --bar=baz
```
And have it all work correctly.

I mean, that's obviously not essential. But given it's not hard to achieve either it seems worthwhile.